### ISSUE TITLE
Transform Summary Relationships Produce Unintuitive Result

### ISSUE THREAD

**Issue by rnugent3** [HydrologicEngineeringCenter/HEC-FDA#1156]
We have a case where mean EAD increases when a regulated-unregulated transform flow function is applied. This phenomenon only occurs when EAD is computed with uncertainty. The preview compute produces the expected result: deterministic EAD is decreased by flow regulation. The study of interest is Sum City, Impact Area 7, DB Base Scenario. 

The behavior appears to be related to the internally computed stage-damage functions. The consequences compute with uncertainty results in significant damage at the most frequent stage. This result points to whacky modeling but a result seen in the field and thus a result we need to be able to handle. 

I created two more sets of stage-damage functions to test my suspicion. Manual stage-damage functions. The first set looks normal, has stages that have zero damage. 
![image](https://github.com/user-attachments/assets/dc6a53f0-dc50-45b5-810a-76a8ce8b7da5)
I computed EAD with uncertainty, without regulation:
![image](https://github.com/user-attachments/assets/e1f12d66-8611-494d-a137-220db529a670)
I computed EAD with uncertainty, with regulation, EAD goes down, the expected behavior: 
![image](https://github.com/user-attachments/assets/159556f4-85f9-4aed-b924-db6c88dd32bf)

The next is a truncated version, there are no stages with zero damage. 
![image](https://github.com/user-attachments/assets/73b55204-7dac-42d7-a552-0c936231ec19)
I computed EAD with uncertainty, without regulation:
![image](https://github.com/user-attachments/assets/7ade25e2-9204-4fae-90c8-c0e5563e3233)
I computed EAD with uncertainty, with regulation, EAD goes up, the *unexpected* behavior: 
![image](https://github.com/user-attachments/assets/26d877f5-479c-428d-9ceb-3246071d2d53)

Somewhere, the math goes awry, and I haven't figured out where that math in the full compute with uncertainty is going wrong, but I know that the stage-damage function is the aggressor. 

A recommended path forward might be to force a zero-damaging stage on the stage-damage function with uncertainty and see if this behavior persists. The trick would be, at what stage would we force zero damage? I think that we could go really close to the most frequent stage, something like 0.1 feet, because we need to put a boundary around the bad modeling.

**Comment by rnugent3** [comment]
Original model is here: `\\share.hec.usace.army.mil\USACE\HEC-FDA\Version_2.0\TechTransfer\ApplicationsGuide\Sum City Model`

**Comment by rnugent3** [comment]
Manual stage damage functions are attached
[sumcitymanualstagedamage.xlsx](https://github.com/user-attachments/files/17426098/sumcitymanualstagedamage.xlsx)

**Comment by rnugent3** [comment]
It looks like function composition is failing.

**Comment by rnugent3** [comment]
This appears to be a binary search failure. In the below screen shot, we're searching for a value of 1196.51 among the Yvals. 1196.51 falls between indices 1 and 2. Binary search returns -3, as if 1196.51 falls between 2 and 3. This explains why composition with reg-unreg produces higher damage. 

![image](https://github.com/user-attachments/assets/c352c3bb-4403-4ec1-a77d-aa606ed0a250)

**Comment by rnugent3** [comment]
The most strange result is that when I apply a regulated-unregulated function where the regulated flows perfectly match the unregulated flows for this particular study, mean EAD changes. 

It is only Sum city that demonstrates this behavior. Applying similar functions to Muncie, such flow transform that demonstrates regulation or a flow transform where there is no transformation so that regulation = unregulated, we get the expected result. So Muncie EAD goes down for regulation and is unchanged when reg = unreg deterministically.

**Comment by rnugent3** [comment]
Ok. I think we have a general transformation problem. 

Three different studies: Sum City, Muncie, and West Sac. Apply flow or stage transform functions where regulated = unregulated or exterior = interior for the entire domain and results change. 

Here is an example of West Sac. 

Original West Sac results. EAD = 1.825M
![image](https://github.com/user-attachments/assets/905c9518-f81d-453c-96e6-0e8038726011)
West Sac with Reg = Unreg. EAD = 1.889M
![image](https://github.com/user-attachments/assets/440b6e2e-18a2-409a-b49b-8949d80807f6)
West Sac with Ext = Int. EAD = 1.816M
![image](https://github.com/user-attachments/assets/79ede5fa-bd71-478c-9cf4-961dfbba4bba)

**Comment by rnugent3** [comment]
Here is Muncie. Similar in that flow transform causes bigger deviation. Different in that flow transform decreases EAD instead of increasing EAD. 

Original. EAD = .986M
![image](https://github.com/user-attachments/assets/d0a51243-5ee4-414b-a8d6-41c95775287c)
Reg = Unreg. EAD = 0.878M
![image](https://github.com/user-attachments/assets/a276f0f8-0a86-460d-ae0c-c21ba138251e)
Int = Ext. EAD = 0.972M
![image](https://github.com/user-attachments/assets/baf89051-d63c-4737-9195-c8e54c26c7cb)

**Comment by rnugent3** [comment]
I can track in the IDE that a flow transformation with `X = Y` does not have any impact on a flow-frequency function, provided that there is sufficient overlap between the two functions.

**Comment by rnugent3** [comment]
This situation continues to defy logic. I have added some diagnostic code to prove that a flow transformation with `X = Y` does not have any impact on a flow-frequency function, provided that there is sufficient overlap between the two functions.

In the below code, I added the check that the transformed flow-frequency function matches the sampled flow-frequency function after composing with a regulated-unregulated function where `X=Y`. 

The functions are always equal, and yet, mean EAD is different. 
```
                                else
                                {
                                    PairedData inflow_outflow_sample = _UnregulatedRegulated.SamplePairedData(threadlocalRandomProvider.NextRandom(), computeIsDeterministic); //should be a random number
                                    PairedData transformff = inflow_outflow_sample.compose(frequencyDischarge);
                                    bool functionsAreEqual = transformff.Equals(frequencyDischarge);
                                    if (!functionsAreEqual)
                                    {
                                        functionsAreEqual = false;
                                        //STOP THE WORLD IS OVER
                                    }
                                    PairedData discharge_stage_sample = _DischargeStage.SamplePairedData(threadlocalRandomProvider.NextRandom(), computeIsDeterministic);//needs to be a random number
                                    PairedData frequency_stage = discharge_stage_sample.compose(transformff);
                                    ComputeFromStageFrequency(threadlocalRandomProvider, frequency_stage, i, computeWithDamage, computeIsDeterministic);
                                }
```

**Comment by rnugent3** [comment]
I changed that code again. I pulled `transformff` out of the compose with the stage-discharge and replaced it with dischargeFrequency, so, in the one place we should be using a transform flow, remove it from the compute in the code, and the transform flow still impacts mean EAD, so it seems like we might have included the transform in the logic somewhere else???

![image](https://github.com/user-attachments/assets/035c38a4-f709-4dfc-a620-3952648a2a8d)

**Comment by rnugent3** [comment]
When I pull out the transformff from the compose, I can see in the preview compute that applying a transform function has no affect on deterministic EAD. 

And still, with flow transformation completely unplugged from the compute, selection of a transform function in the construction of an impact area scenario causes a different mean EAD.

I can also show that the difference in mean EAD is insensitive to the values in the assigned transformation function. So I've pulled transformation out of the compute, I assign any transformation function to the impact area scenario, and mean EAD is 214k, independent of whatever flow values are in the function. If I don't assign a transformation function, mean  EAD is 190k.

**Comment by rnugent3** [comment]
Could it be the sequence of random numbers that matters here?

**Comment by rnugent3** [comment]
Yes. Random number sequence, YIKES!

**Comment by rnugent3** [comment]
Currently, we use one sequence of random numbers for all summary relationship sampling in an EAD compute with uncertainty. 

The EAD distribution appears to be very sensitive to the sequence of "random" numbers. We knew there was sensitivity, which is why the compute needs to be seeded. 

What we missed is that the seeding of the compute breaks when summary relationships are added. Adding the sampling regulated-unregulated transform flow function means that the sequence of random numbers is shifted, and now a different set of random numbers are injected in the sampling of the stage-discharge function, which causes the EAD distribution to be different. 

What I think this means is that every single summary relationship needs its own seed and its own sequence of random numbers, so that the sequence of random numbers that a summary relationship accepts in sampling is the same for every compute.

**Comment by rnugent3** [comment]
And for that matter, we'll also need separately seeded random number generators for each stage-damage function, where there is one function for each dam cat - asset cat combination, and there are separate stage-damage functions for nonfail when fail and nonfail are being combined for total risk.

**Comment by rnugent3** [comment]
> We have a case where mean EAD increases when a regulated-unregulated transform flow function is applied. This phenomenon only occurs when EAD is computed with uncertainty. The preview compute produces the expected result: deterministic EAD is decreased by flow regulation. The study of interest is Sum City, Impact Area 7, DB Base Scenario.
> 
> The behavior appears to be related to the internally computed stage-damage functions. The consequences compute with uncertainty results in significant damage at the most frequent stage. This result points to whacky modeling but a result seen in the field and thus a result we need to be able to handle.
> 
> I created two more sets of stage-damage functions to test my suspicion. Manual stage-damage functions. The first set looks normal, has stages that have zero damage. ![image](https://private-user-images.githubusercontent.com/25163165/377671906-dc6a53f0-dc50-45b5-810a-76a8ce8b7da5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0NTQyOTksIm5iZiI6MTczMTQ1Mzk5OSwicGF0aCI6Ii8yNTE2MzE2NS8zNzc2NzE5MDYtZGM2YTUzZjAtZGM1MC00NWI1LTgxMGEtNzZhOGNlOGI3ZGE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEyVDIzMjYzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRkZmU3OGQ1Y2VlNzFlNWEyODJkODJiN2IwNmRjYjY0ODA4Y2ZiYTA5MDcwOGYyMzRkMjI2NDk2MDMwZmY0NDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0VCf9sNvUKX6CjKliSX26kCZ30Pd0EE2hM7ZE_q2FLQ) I computed EAD with uncertainty, without regulation: ![image](https://private-user-images.githubusercontent.com/25163165/377672564-e1f12d66-8611-494d-a137-220db529a670.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0NTQyOTksIm5iZiI6MTczMTQ1Mzk5OSwicGF0aCI6Ii8yNTE2MzE2NS8zNzc2NzI1NjQtZTFmMTJkNjYtODYxMS00OTRkLWExMzctMjIwZGI1MjlhNjcwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEyVDIzMjYzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc4OTJiM2ZlODYzMTkyNWE4ZTBlZjk5MGQ3ZTlhOWZjMDljNTIwYmNiM2QwMzFmOTRjOTY3ZGUyOTQyYjRhODYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6sXsLe_Hi2iFno-0Bw08dgzEaETJ9eMtUugevg4MX1s) I computed EAD with uncertainty, with regulation, EAD goes down, the expected behavior: ![image](https://private-user-images.githubusercontent.com/25163165/377672535-159556f4-85f9-4aed-b924-db6c88dd32bf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0NTQyOTksIm5iZiI6MTczMTQ1Mzk5OSwicGF0aCI6Ii8yNTE2MzE2NS8zNzc2NzI1MzUtMTU5NTU2ZjQtODVmOS00YWVkLWI5MjQtZGI2Yzg4ZGQzMmJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEyVDIzMjYzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM5YWQ1YzViNDBhYjRhNWE3MzExZmI1ZWIzM2M5NTk5ZWE4ZmI5MDI4ZDM5MzMyNDEwMzliMzg4NjhmYzNiZTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AnhaDsy78gB7J7U6BVDZAf4DHSXhzcuybLH_Uycd-q8)
> 
> The next is a truncated version, there are no stages with zero damage. ![image](https://private-user-images.githubusercontent.com/25163165/377672097-73b55204-7dac-42d7-a552-0c936231ec19.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0NTQyOTksIm5iZiI6MTczMTQ1Mzk5OSwicGF0aCI6Ii8yNTE2MzE2NS8zNzc2NzIwOTctNzNiNTUyMDQtN2RhYy00MmQ3LWE1NTItMGM5MzYyMzFlYzE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEyVDIzMjYzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3Mzg5ZGYwYWNkZjgwNzA0NGM4NTBhYTIzZDEzNWFhZTUxMmY5NWM1NGVmN2MwMDFiZTkxNGVlMTNlOGQzNzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.nU7e809EkbiiSuTyTOle9P3eqNs4mAy-PcUTOibRNhA) I computed EAD with uncertainty, without regulation: ![image](https://private-user-images.githubusercontent.com/25163165/377672780-7ade25e2-9204-4fae-90c8-c0e5563e3233.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0NTQyOTksIm5iZiI6MTczMTQ1Mzk5OSwicGF0aCI6Ii8yNTE2MzE2NS8zNzc2NzI3ODAtN2FkZTI1ZTItOTIwNC00ZmFlLTkwYzgtYzBlNTU2M2UzMjMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEyVDIzMjYzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThlOTU2ZWYxM2RkMDYwZjliYTk5MjYyNzY1YzMzYTRmNjU4MjcyZTg1ODM3MWJhNDg4OThmOGI2NzZlNmExMmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZvQ1H3-Kzuumfwv5tpUUxBTRJe35T2UZ1cCaHKj7qPs) I computed EAD with uncertainty, with regulation, EAD goes up, the _unexpected_ behavior: ![image](https://private-user-images.githubusercontent.com/25163165/377672744-26d877f5-479c-428d-9ceb-3246071d2d53.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0NTQyOTksIm5iZiI6MTczMTQ1Mzk5OSwicGF0aCI6Ii8yNTE2MzE2NS8zNzc2NzI3NDQtMjZkODc3ZjUtNDc5Yy00MjhkLTljZWItMzI0NjA3MWQyZDUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEyVDIzMjYzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJlOTQ5NjFmZTcyZjQxYzFiYTc1ODM4ZTQxMTAwNzc2OGVmZDhkMDczYjRmMDYxYTFhN2E2ODZhZGVlNmZkN2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.DlksuGH3upDaF2Gyhd3sepFv4uiXOD-pRb95vd2ANzk)
> 
> Somewhere, the math goes awry, and I haven't figured out where that math in the full compute with uncertainty is going wrong, but I know that the stage-damage function is the aggressor.
> 
> A recommended path forward might be to force a zero-damaging stage on the stage-damage function with uncertainty and see if this behavior persists. The trick would be, at what stage would we force zero damage? I think that we could go really close to the most frequent stage, something like 0.1 feet, because we need to put a boundary around the bad modeling.

I am not tracking how the random number issue relates to the differences observed when damage is truncated versus not as in what I showed at the top of this ticket. I'll need to look at that after. 

First, if I apply a transform function where there is no transformation so that `X=Y` then there should be no difference in the result and I think we achieve that through enhanced random number generation.

**Comment by rnugent3** [comment]
I am thinking about each UPD having its own set of seeded RNG by iteration and passing the iteration into sample paired data to use the right RNG during the compute. 

All of the UPD can be given the same set of master seeds created for the provided convergence criteria. We can construct the set of RNG on command, and not worry about reading and writing to file. 

Sample paired data can be overloaded to accept and use indices, the other sample paired data can remain as is. 

It is possible that there are similar ramifications in the consequences engine. If we add an occupancy type to an occupancy type set, could that break the sequence of seeded random numbers, resulting in different damage distributions? Eek

**Comment by rnugent3** [comment]
Generally speaking, I think that every single THING that can be sampled, e.g. UPD, Value Uncertainty, Value Ratio Uncertainty, Continuous Dist, Structure, they *all* have to own their own RNGs so that their sampling can be replicated independent of adding or removing other things.

**Comment by rnugent3** [comment]
> I am thinking about each UPD having its own set of seeded RNG by iteration and passing the iteration into sample paired data to use the right RNG during the compute. 
> 
> All of the UPD can be given the same set of master seeds created for the provided convergence criteria. We can construct the set of RNG on command, and not worry about reading and writing to file. 
> 
> Sample paired data can be overloaded to accept and use indices, the other sample paired data can remain as is. 
> 
> It is possible that there are similar ramifications in the consequences engine. If we add an occupancy type to an occupancy type set, could that break the sequence of seeded random numbers, resulting in different damage distributions? Eek

As each thing might have its own RNG, we can seed it uniquely and tell it to populate an array of random numbers and then during the compute reference the correct random number for iteration i so that we control for random order of iterations in a parallel compute

**Comment by rnugent3** [comment]
Closed with PR #1168
